// nw-toast/src/toast.js
//@flow
import React, {Component} from "react";
import styles from './style.module.css';

export default class Toast extends Component {
    state: {
        text: string,
        isShow: boolean
    };

    _handle: ?number;

    constructor () {
        super();
        this.state = {
            text: "",
            isShow: false
        };
    }

    toast ({text, delay = 2000}: {text: string, delay: number}) {
        if (this._handle !== undefined) {
            window.clearTimeout(this._handle);
        }

        this.setState({
            text,
            isShow: true
        });

        this._handle = window.setTimeout(() => {
            this.setState({
                isShow: false
            })
        }, delay);
    }

    componentDidMount() {
        this.toast(this.props);
    }

    render () {
        const coverClassName = styles.cover + ' ' + styles.aBouncein;

        return (
            <div className={coverClassName} style={{"display": this.state.isShow? "table" : "none"}}>
                <div className={styles.inner}>
                    <div className={styles.main}>{this.state.text}</div>
                </div>
            </div>
        );
    }
}
