// nw-share/webpack.config.js
var path = require('path');
const webpack = require('webpack');

var webpackConfig = {
  entry: {
      index: './src/index.js'
  },
  output: {
    path: __dirname + '/dist',
    filename: 'share-dist.js',
    library: 'Share',
    libraryTarget: 'umd',
    chunkFilename: '[name]_bundle.js',
    publicPath: 'https://easyread.nosdn.127.net/components/share4/'
  },
  plugins: [

    new webpack.DefinePlugin({
        ENV: {
            name: '"production"'
        }
    }),

    // new webpack.optimize.UglifyJsPlugin({
    //     compress: {
    //         warnings: false
    //     },
    //     beautify: false,
    //     sourceMap: false
    // }),
],

  module: {

    rules: [
        {
            test: /\.js$/,
            exclude: /node_modules/,
            use: 'babel-loader'
        },

        {
            test: /\.css$/,
            use: [
                {
                    loader: 'style-loader'
                },
                {
                    loader: 'css-loader'
                }
            ]
        },

        {test: /\.(png|jpg|TTF|eot)$/, use: "url-loader?prefix=res/&limit=5000"}
    ]
    }
};

module.exports = webpackConfig;
