// nw-share/es/wechat/index.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

var Weixin = /*#__PURE__*/function () {
  function Weixin(props) {
    _classCallCheck(this, Weixin);

    this.setData(props);
    this.__JSBridgeName = props.JSBridgeName;
    var JSBridge = window[this.__JSBridgeName];

    if (JSBridge) {
      this.__onBridgeReady();
    } else {
      document.addEventListener(this.__JSBridgeName + 'Ready', this.__onBridgeReady.bind(this), false);
    } // 易信分享到微信只是通过`window.shareData`设置分享参数


    if (this.__JSBridgeName === 'YixinJSBridge') {
      var shareData = props.data;
      window.shareData = {
        "imgUrl": shareData.picurl,
        "tImgUrl": shareData.picurl,
        "fImgUrl": shareData.picurl,
        "wImgUrl": shareData.picurl,
        "timeLineLink": shareData.link,
        "sendFriendLink": shareData.link,
        "weiboLink": shareData.link,
        "tTitle": shareData.title,
        "tContent": shareData.description,
        "fTitle": shareData.title,
        "fContent": shareData.description,
        "wContent": shareData.text
      };
    }
  }

  _createClass(Weixin, [{
    key: "setData",
    value: function setData(_ref) {
      var data = _ref.data,
          callback = _ref.callback;

      if (data) {
        this.__data = {
          link: data.link,
          title: data.title,
          img_url: data.picurl,
          desc: data.description,
          intro: data.text
        };
      }

      if (callback) {
        this.__callback = callback;
      }
    }
  }, {
    key: "__onBridgeReady",
    value: function __onBridgeReady() {
      this.__JSBridge = window[this.__JSBridgeName];

      this.__JSBridge.on('menu:share:appmessage', this.__shareFriend.bind(this));

      this.__JSBridge.on('menu:share:timeline', this.__shareTimeline.bind(this));

      this.__JSBridge.on('menu:share:weibo', this.__shareWeibo.bind(this));
    } // http://jira.netease.com/browse/WYMH-4494

  }, {
    key: "__getLink",
    value: function __getLink() {
      var link = this.__data.link;
      var hash = '';
      var hashIndex = link.indexOf('#');

      if (hashIndex !== -1) {
        hash = link.slice(hashIndex);
        link = link.slice(0, hashIndex);
      }

      if (link.indexOf('?') !== -1) {
        link += '&fromwechat=1';
      } else {
        link += '?fromwechat=1';
      }

      link += hash;
      return link;
    }
  }, {
    key: "__shareWeibo",
    value: function __shareWeibo() {
      var _this = this;

      this.__JSBridge.invoke('shareWeibo', {
        "content": this.__data.title,
        "url": this.__getLink()
      }, function (res) {
        // _report('weibo', res.err_msg);
        _this.__callback && _this.__callback({
          code: 0,
          platform: "".concat(_this.__JSBridgeName, "_weibo")
        });
      });
    }
  }, {
    key: "__shareTimeline",
    value: function __shareTimeline() {
      var _this2 = this;

      this.__JSBridge.invoke('shareTimeline', {
        "img_url": this.__data.img_url,
        "link": this.__getLink(),
        "desc": this.__data.desc,
        "title": this.__data.title,
        "img_width": "640",
        "img_height": "640"
      }, function (res) {
        // _report('timeline', res.err_msg);
        // share_timeline:ok
        _this2.__callback && _this2.__callback({
          code: 0,
          platform: "".concat(_this2.__JSBridgeName, "_timeline")
        });
      });
    }
  }, {
    key: "__shareFriend",
    value: function __shareFriend() {
      var _this3 = this;

      this.__JSBridge.invoke('sendAppMessage', {
        "appid": this.__data.appId,
        "img_url": this.__data.img_url,
        "link": this.__getLink(),
        "desc": this.__data.desc,
        "title": this.__data.title,
        "img_width": "640",
        "img_height": "640"
      }, function (res) {
        // _report('send_msg', res.err_msg);
        // send_app_msg:cancel
        // send_app_msg:confirm
        _this3.__callback && _this3.__callback({
          code: 0,
          platform: "".concat(_this3.__JSBridgeName, "_friend")
        }); // favorite
        // send_app_msg:ok
      });
    }
  }]);

  return Weixin;
}();

export { Weixin as default };
