// nw-share/es/wechat/lofter.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

// lofter域名的网页在微信内的分享
import axios from 'axios';
var wxSdkUrl = "https://res.wx.qq.com/open/js/jweixin-1.0.0.js";

function loadScript(url) {
  return new Promise(function (resolve, reject) {
    if (window.wx && typeof window.wx.config === 'function') {
      resolve();
      return;
    }

    var script = document.createElement('script');
    script.src = url;

    script.onload = function () {
      resolve();
    };

    script.onerror = function () {
      reject('sdk load faied');
    };

    document.body.appendChild(script);
  });
}

export function shareLofterInWechat(shareConfig, callback) {
  var scriptPromise = loadScript(wxSdkUrl); // 20211210修改为www.lofter.com域名，该接口修改为支持跨域

  var dataPromise = axios.get('//www.lofter.com/spread/inherit/track/getWeiXinOpenSignature', {
    method: 'get',
    type: 'json',
    params: {
      url: location.href.split('#')[0]
    }
  });
  Promise.all([scriptPromise, dataPromise]).then(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
        res = _ref2[1];

    var data = res.data;

    if (!!data && data.code == 200) {
      window.wx.config({
        debug: false,
        appId: 'wxd8296e70dd0d29c3',
        timestamp: data.data.timestamp,
        nonceStr: data.data.noncestr,
        signature: data.data.signature,
        jsApiList: ['onMenuShareTimeline', 'onMenuShareAppMessage']
      });
      window.wx.ready(function () {
        wx.onMenuShareTimeline({
          title: shareConfig.title,
          link: shareConfig.link,
          imgUrl: shareConfig.picurl,
          success: callback
        });
        window.wx.onMenuShareAppMessage({
          title: shareConfig.title,
          desc: shareConfig.description,
          link: shareConfig.link,
          imgUrl: shareConfig.picurl,
          success: callback
        });
      });
    }
  }, function (_ref3) {
    var _ref4 = _slicedToArray(_ref3, 2),
        scriptErr = _ref4[0],
        dataErr = _ref4[1];

    if (scriptErr) {
      console.error(scriptErr);
    }

    ;

    if (dataErr) {
      console.error(dataErr);
    }
  });
}
