// nw-share/src/browser-share.js
import queryString from 'qs';
import copy from 'copy-to-clipboard';

const appendStyle = (css) => {
    const head = document.head || document.getElementsByTagName('head')[0],
    style = document.createElement('style');

    style.type = 'text/css';
    if (style.styleSheet){
        style.styleSheet.cssText = css;
    } else {
        style.appendChild(document.createTextNode(css));
    }

    head.appendChild(style);
}

const onDialogClick = (event) => {
    const shareData = localShareData;

    var rect = dialog.getBoundingClientRect();
    var isInDialog=(rect.top <= event.clientY && event.clientY <= rect.top + rect.height
      && rect.left <= event.clientX && event.clientX <= rect.left + rect.width);
    if (!isInDialog) {
        close();
    }

    event.preventDefault();

    const target = event.target;

    switch (target.dataset.type) {

        case 'wechat':
            const text = `${shareData.link} ${shareData.title}`;
            const isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);

            if (isSafari) {
                const success = copy(text, {
                    message: '请手动复制以下内容分享给好友'
                });

                if (success) {
                    alert('已复制分享内容，请打开微信分享给好友');
                }
            } else {
                window.prompt('请手动复制以下内容分享给好友', text);
            }

            close();
        break;

        case 'qzone':
            window.location.href = getQzoneUrl(shareData);
            break;

        case 'weibo':
            window.location.href = getWeiboUrl(shareData);
            break;
    }
};

const close = () => {
    if (dialog.hasAttribute('open')) {
        dialog.close();
    }

    localShareData = null;
}

const onBackdropClick = (event) => {
    event.stopPropagation();

    close();
}


var css = `
    .nw-share {
        border: 1px solid #999;
        border-radius: 4px;
        white-space: nowrap;
        width: 177px;
        height: 82px;
        box-sizing: border-box;
        padding: 16px;
    }
    .nw-share::backdrop {
        background-color: rgba(0,0,0,.5);
    }
    .nw-share + .backdrop {
        background-color: rgba(0,0,0,.5);
    }
    .nw-share__channel {
        background-image: url("https://yuedust.yuedu.126.net/comic_static/web/images/sprites/sns@2x.png");
        background-repeat: no-repeat;
        width: 20px;
        height: 20px;
        display: inline-block;
        background-size: 85px 79px;
        margin: 10px;
    }
    .nw-share__channel:focus {
        outline: none;
    }
    .nw-share__channel--wechat {
        background-position: -28px -29px;
        width: 23px;
        height: 20px;
    }
    .nw-share__channel--qqzone {
        background-position: -2px -2px;
        width: 26px;
        height: 23px;
    }
    .nw-share__channel--weibo {
        background-position: -32px -2px;
        width: 24px;
        height: 20px;
    }
`;

appendStyle(css);

const getQzoneUrl = (data) => {
    var shareData = {
        url: data.link,
        showcount: '1',
        /*是否显示分享总数,显示：'1'，不显示：'0' */
        /*  desc: "//正在看《" + tmpData.title + "》。" + tmpData.summary, /!*默认分享理由(可选)*!/*/
        summary: data.description,
        /*分享摘要(可选)*/
        title: data.title,
        /*分享标题(可选)*/
        /*site: '网易漫画',*/
        /*分享来源 如：腾讯网(可选)*/
        pics: data.picurl,
        /*分享图片的路径(可选)*/
        style: '201',
        width: 32,
        height: 32
    };

    return 'http://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?' + queryString.stringify(shareData);
}

const getWeiboUrl = (data) => {
    const shareData = {
        appkey: '3254753781',
        url: data.link,
        title: data.text,
        pic: data.picurl 
    };
    return 'http://service.weibo.com/share/share.php?' + queryString.stringify(shareData);
}

let dialog;
let localShareData;
let loadingPolyfill = false;

export const browserShare = (shareData) => {

    localShareData = shareData;

    if (dialog === undefined) {
        dialog = document.createElement('dialog');
        dialog.className = 'nw-share fixed';
    
        dialog.addEventListener('click', onDialogClick, false);

        document.body.appendChild(dialog);

        const html = `
            <a href="#" data-type="wechat" target="_blank" class="nw-share__channel nw-share__channel--wechat"></a>
            <a href="#" data-type="weibo" target="_blank" class="nw-share__channel nw-share__channel--weibo"></a>
            <a href="#" data-type="qzone" target="_blank" class="nw-share__channel nw-share__channel--qqzone"></a>
        `;

        dialog.innerHTML = html;

        if (typeof HTMLDialogElement !== 'function') {
            loadingPolyfill = true;
            import('./dialog-polyfill').then((dialogPolyfill) => {
                loadingPolyfill = false;
                dialogPolyfill.default.registerDialog(dialog);

                dialog.showModal();
                dialog.nextElementSibling.addEventListener('touchstart', onBackdropClick, false);
            })
            .catch((e) => {
                alert('分享组件载入失败，请重试');
                unmount();
                loadingPolyfill = false;
            });

            return;
        }
    } else if (loadingPolyfill) {
        return;
    }

    dialog.showModal();
}


export const unmount = () => {
    if (!dialog) return;

    dialog.removeEventListener('click', onDialogClick, false);

    if (typeof HTMLDialogElement !== 'function'
        && dialog.nextElementSibling.classList.contains('backdrop')) {
        dialog.nextElementSibling.removeEventListener('touchstart', onBackdropClick, false);
    }

    close();
    dialog.remove();
    dialog = null;
}
