// nw-share/es/browser-share.js
import queryString from 'qs';
import copy from 'copy-to-clipboard';

var appendStyle = function appendStyle(css) {
  var head = document.head || document.getElementsByTagName('head')[0],
      style = document.createElement('style');
  style.type = 'text/css';

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }

  head.appendChild(style);
};

var onDialogClick = function onDialogClick(event) {
  var shareData = localShareData;
  var rect = dialog.getBoundingClientRect();
  var isInDialog = rect.top <= event.clientY && event.clientY <= rect.top + rect.height && rect.left <= event.clientX && event.clientX <= rect.left + rect.width;

  if (!isInDialog) {
    close();
  }

  event.preventDefault();
  var target = event.target;

  switch (target.dataset.type) {
    case 'wechat':
      var text = "".concat(shareData.link, " ").concat(shareData.title);
      var isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);

      if (isSafari) {
        var success = copy(text, {
          message: '请手动复制以下内容分享给好友'
        });

        if (success) {
          alert('已复制分享内容，请打开微信分享给好友');
        }
      } else {
        window.prompt('请手动复制以下内容分享给好友', text);
      }

      close();
      break;

    case 'qzone':
      window.location.href = getQzoneUrl(shareData);
      break;

    case 'weibo':
      window.location.href = getWeiboUrl(shareData);
      break;
  }
};

var close = function close() {
  if (dialog.hasAttribute('open')) {
    dialog.close();
  }

  localShareData = null;
};

var onBackdropClick = function onBackdropClick(event) {
  event.stopPropagation();
  close();
};

var css = "\n    .nw-share {\n        border: 1px solid #999;\n        border-radius: 4px;\n        white-space: nowrap;\n        width: 177px;\n        height: 82px;\n        box-sizing: border-box;\n        padding: 16px;\n    }\n    .nw-share::backdrop {\n        background-color: rgba(0,0,0,.5);\n    }\n    .nw-share + .backdrop {\n        background-color: rgba(0,0,0,.5);\n    }\n    .nw-share__channel {\n        background-image: url(\"https://yuedust.yuedu.126.net/comic_static/web/images/sprites/sns@2x.png\");\n        background-repeat: no-repeat;\n        width: 20px;\n        height: 20px;\n        display: inline-block;\n        background-size: 85px 79px;\n        margin: 10px;\n    }\n    .nw-share__channel:focus {\n        outline: none;\n    }\n    .nw-share__channel--wechat {\n        background-position: -28px -29px;\n        width: 23px;\n        height: 20px;\n    }\n    .nw-share__channel--qqzone {\n        background-position: -2px -2px;\n        width: 26px;\n        height: 23px;\n    }\n    .nw-share__channel--weibo {\n        background-position: -32px -2px;\n        width: 24px;\n        height: 20px;\n    }\n";
appendStyle(css);

var getQzoneUrl = function getQzoneUrl(data) {
  var shareData = {
    url: data.link,
    showcount: '1',

    /*是否显示分享总数,显示：'1'，不显示：'0' */

    /*  desc: "//正在看《" + tmpData.title + "》。" + tmpData.summary, /!*默认分享理由(可选)*!/*/
    summary: data.description,

    /*分享摘要(可选)*/
    title: data.title,

    /*分享标题(可选)*/

    /*site: '网易漫画',*/

    /*分享来源 如：腾讯网(可选)*/
    pics: data.picurl,

    /*分享图片的路径(可选)*/
    style: '201',
    width: 32,
    height: 32
  };
  return 'http://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?' + queryString.stringify(shareData);
};

var getWeiboUrl = function getWeiboUrl(data) {
  var shareData = {
    appkey: '3254753781',
    url: data.link,
    title: data.text,
    pic: data.picurl
  };
  return 'http://service.weibo.com/share/share.php?' + queryString.stringify(shareData);
};

var dialog;
var localShareData;
var loadingPolyfill = false;
export var browserShare = function browserShare(shareData) {
  localShareData = shareData;

  if (dialog === undefined) {
    dialog = document.createElement('dialog');
    dialog.className = 'nw-share fixed';
    dialog.addEventListener('click', onDialogClick, false);
    document.body.appendChild(dialog);
    var html = "\n            <a href=\"#\" data-type=\"wechat\" target=\"_blank\" class=\"nw-share__channel nw-share__channel--wechat\"></a>\n            <a href=\"#\" data-type=\"weibo\" target=\"_blank\" class=\"nw-share__channel nw-share__channel--weibo\"></a>\n            <a href=\"#\" data-type=\"qzone\" target=\"_blank\" class=\"nw-share__channel nw-share__channel--qqzone\"></a>\n        ";
    dialog.innerHTML = html;

    if (typeof HTMLDialogElement !== 'function') {
      loadingPolyfill = true;
      import('./dialog-polyfill').then(function (dialogPolyfill) {
        loadingPolyfill = false;
        dialogPolyfill["default"].registerDialog(dialog);
        dialog.showModal();
        dialog.nextElementSibling.addEventListener('touchstart', onBackdropClick, false);
      })["catch"](function (e) {
        alert('分享组件载入失败，请重试');
        unmount();
        loadingPolyfill = false;
      });
      return;
    }
  } else if (loadingPolyfill) {
    return;
  }

  dialog.showModal();
};
export var unmount = function unmount() {
  if (!dialog) return;
  dialog.removeEventListener('click', onDialogClick, false);

  if (typeof HTMLDialogElement !== 'function' && dialog.nextElementSibling.classList.contains('backdrop')) {
    dialog.nextElementSibling.removeEventListener('touchstart', onBackdropClick, false);
  }

  close();
  dialog.remove();
  dialog = null;
};
