// nw-share/src/wechat/index.js

export default class Weixin {
	constructor(props) {
      this.setData(props);
  		this.__JSBridgeName = props.JSBridgeName;
  		const JSBridge = window[this.__JSBridgeName];

		if (JSBridge) {
			this.__onBridgeReady();
		} else {
			document.addEventListener(this.__JSBridgeName + 'Ready', this.__onBridgeReady.bind(this), false);
        }
            
        // 易信分享到微信只是通过`window.shareData`设置分享参数
        if (this.__JSBridgeName === 'YixinJSBridge') {
            const shareData = props.data;
            window.shareData = {
                "imgUrl": shareData.picurl,
                "tImgUrl": shareData.picurl,
                "fImgUrl": shareData.picurl,
                "wImgUrl": shareData.picurl,
                "timeLineLink": shareData.link,
                "sendFriendLink": shareData.link,
                "weiboLink": shareData.link,
                "tTitle": shareData.title,
                "tContent": shareData.description,
                "fTitle": shareData.title,
                "fContent": shareData.description,
                "wContent": shareData.text
            };
        }
	}

  setData({data, callback}) {
    if (data) {
        this.__data = {
            link: data.link,
            title: data.title,
            img_url: data.picurl,
            desc: data.description,
            intro: data.text
        };
    }
    if (callback) {
        this.__callback = callback;
    }
  }

	__onBridgeReady() {
        this.__JSBridge = window[this.__JSBridgeName];

        this.__JSBridge.on('menu:share:appmessage', this.__shareFriend.bind(this));
        this.__JSBridge.on('menu:share:timeline', this.__shareTimeline.bind(this));
        this.__JSBridge.on('menu:share:weibo', this.__shareWeibo.bind(this));
    }

    // http://jira.netease.com/browse/WYMH-4494
    __getLink() {
        let link = this.__data.link;
        let hash = '';
        const hashIndex = link.indexOf('#')

        if (hashIndex !== -1) {
            hash = link.slice(hashIndex);
            link = link.slice(0, hashIndex);
        }
        
        if (link.indexOf('?') !== -1) {
            link += '&fromwechat=1';
        } else {
            link += '?fromwechat=1';
        }

        link += hash;

        return link;
    }

	__shareWeibo() {
      this.__JSBridge.invoke('shareWeibo', {
          "content": this.__data.title,
          "url": this.__getLink()
      }, (res) => {
          // _report('weibo', res.err_msg);

          this.__callback && this.__callback({
            code: 0,
            platform: `${this.__JSBridgeName}_weibo`
          });
      });
    }

    __shareTimeline() {
       this.__JSBridge.invoke('shareTimeline', {
            "img_url": this.__data.img_url,
            "link": this.__getLink(),
            "desc": this.__data.desc,
            "title": this.__data.title,
            "img_width":"640",
            "img_height":"640"
        }, (res) => {
            // _report('timeline', res.err_msg);
          // share_timeline:ok
          this.__callback && this.__callback({
              code: 0,
              platform: `${this.__JSBridgeName}_timeline`
          });
        });
    }

    __shareFriend() {
      this.__JSBridge.invoke('sendAppMessage', {
          "appid": this.__data.appId,
          "img_url": this.__data.img_url,
          "link": this.__getLink(),
          "desc": this.__data.desc,
          "title": this.__data.title,
          "img_width":"640",
          "img_height":"640"
      }, (res) => {
          // _report('send_msg', res.err_msg);
          // send_app_msg:cancel
          // send_app_msg:confirm

          this.__callback && this.__callback({
            code: 0,
            platform: `${this.__JSBridgeName}_friend`
          });

          // favorite
          // send_app_msg:ok
      });
  }
}
