// nw-share/src/wechat/lofter.js
// lofter域名的网页在微信内的分享
import axios from 'axios';
const wxSdkUrl = `https://res.wx.qq.com/open/js/jweixin-1.0.0.js`;

function loadScript(url) {
  return new Promise( (resolve, reject) => {
    if(window.wx && typeof window.wx.config === 'function'){
      resolve();
      return;
    }
    const script = document.createElement('script');
    script.src = url;
    script.onload = () => {
      resolve();
    };
    script.onerror = () => {
      reject('sdk load faied');
    }
    document.body.appendChild(script);
  })
}

export function shareLofterInWechat(shareConfig, callback) {
  const scriptPromise = loadScript(wxSdkUrl);
  // 20211210修改为www.lofter.com域名，该接口修改为支持跨域
  const dataPromise = axios.get('//www.lofter.com/spread/inherit/track/getWeiXinOpenSignature', {
    method: 'get',
    type: 'json',
    params: {
      url: location.href.split('#')[0],
    },
  });
  Promise.all([scriptPromise, dataPromise]).then(([,res]) => {
    const data = res.data;
    if (!!data && data.code == 200) {
      window.wx.config({
        debug: false,
        appId: 'wxd8296e70dd0d29c3',
        timestamp: data.data.timestamp,
        nonceStr: data.data.noncestr,
        signature: data.data.signature,
        jsApiList: ['onMenuShareTimeline', 'onMenuShareAppMessage'],
      });
      window.wx.ready(() => {
        wx.onMenuShareTimeline({
          title: shareConfig.title,
          link: shareConfig.link,
          imgUrl: shareConfig.picurl,
          success: callback,
        });
        window.wx.onMenuShareAppMessage({
          title: shareConfig.title,
          desc: shareConfig.description,
          link: shareConfig.link,
          imgUrl: shareConfig.picurl,
          success: callback,
        });
      });
    }
  }, ([scriptErr, dataErr]) => {
    if(scriptErr){
      console.error(scriptErr);
    };
    if(dataErr){
      console.error(dataErr);
    }
  })
}
